rfcs/0004-persistent-ledger-builder-controller.md

```diff
@@ -0,0 +1,67 @@
+
+The module type for the sparse diff could be:
+
+```ocaml
+  module type SparseDiff = sig
+    type t
+	type key
+	type value
+
+    type db = Database_intf.S.t
+
+    (* group identity, operations *)
+    val empty : t
+    val add : t -> t -> t
+	val inverse : t -> t
+
+    val insert : key -> value -> t -> t
+	val remove : key -> t -> t
+    val from_list : (key * value) list -> t
+
+    (* look up in diff; if key not present, delegate to database for locked tip *)
+    val get : db -> key -> value option
+  end
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The drawback of this is that the persistent Merkle database is not easy to copy. However, this is by design, as there is only one copy that should be persisted as the central source of truth. Any other information which we wish to persist should live outside of this representation. For instance, if we want to persist the best tip, then only the diff between that tip's database and the locked tip's database should be stored. This is much more efficient for both storage and quick synchronizing, in the case of short term downtimes.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The alternative design to this is to restructure the persistent database so that it can be cheaply copied. This can be done by not keying nodes by their path in the Merkle tree and instead key nodes uniquely and only copy the node that are modified for the new copy. However, there are a few issues with this. For one, it makes many of the common queries to the data structure far less efficient. If you want to find a leaf at the bottom of the tree given a path, you would need to make `k` successive requests to the database, where `k` is the depth of your tree, versus the single request of bulk lookups required by the other representation. This can be allieviated with more indices into the database, but those in turn also introduce more request requirements, and also increase the cost of garbage collection. The second reason is that garbage collection could prove challenging. The main logic behind when something is considered garbage is simple: keep a ref count for each node, and when it's zero, delete it. But, that adds cost to each operation on nodes; or, alternatively, requires us to perform collection cycles in our scheduler. This also becomes exponentially more complex as we add in indices to make the representation more efficient, which will be necessary. Thirdly, this representation adds even more accesses to the database, since you need to do all work with ledger through the database. This will make caching more important for efficiency, and if you are already going to represent the active part of the ledger in memory via a cache, why not just opt for the more efficient representation on the persistent side and use an in memory abstraction for representing the active part of a ledger.
```

Comment:
Sorry, I don't understand the second sentence here. Does that mean using account public keys instead of locations as the database key? 

"alleviated" is misspelled.

Reply:
> Does that mean using account public keys instead of locations as the database key?

I believe so, yes